src/store/modules/conversations.js

```diff
@@ -1,243 +1,119 @@
+      dispatch('send', state.activeConversationId, message)
+    },
 
-  async fetchMessages ({ commit }, conversationId) {
-    commit(types.REQUEST_MESSAGES, { conversationId })
-    let data
-    try {
-      data = await messageAPI.list(conversationId)
-    }
-    catch (error) {
-      commit(types.RECEIVE_MESSAGES_ERROR, { conversationId, error })
-      return
-    }
-    commit(types.RECEIVE_MESSAGES, { conversationId, messages: data.results, cursor: data.next })
-  },
+    async fetchMoreForActiveConversation ({ state, dispatch }) {
+      const conversationId = state.activeConversationId
+      const currentCursor = state.cursors[conversationId]
+      if (!currentCursor) {
+        return
+      }
+      await dispatch('fetchMore', conversationId)
+    },
 
-  async fetchMoreMessages ({ state, commit }) {
-    // fetch more messages for active conversation
-    const conversationId = state.activeConversationId
-    const currentCursor = state.cursors[conversationId]
-    if (!currentCursor) {
-      return
-    }
-    commit(types.REQUEST_MORE_MESSAGES, { conversationId })
+    async setActive ({ commit, dispatch }, conversation) {
+      commit('setActive', { conversationId: conversation.id })
+      commit('setConversation', { conversation })
+      await dispatch('fetch', conversation.id)
+    },
 
-    let data
-    try {
-      data = await messageAPI.listMore(currentCursor)
-    }
-    catch (error) {
-      commit(types.RECEIVE_MORE_MESSAGES_ERROR, { conversationId, error })
-      return
-    }
-    commit(types.RECEIVE_MORE_MESSAGES, { conversationId, messages: data.results, cursor: data.next })
-  },
-}
+    async clearActive ({ commit }) {
+      commit('clearActive')
+    },
 
-export const mutations = {
-  [types.SET_ACTIVE] (state, { conversationId }) {
-    state.activeConversationId = conversationId
-  },
-  [types.CLEAR_ACTIVE] (state) {
-    state.activeConversationId = null
-  },
+    async receiveMessage ({ commit, state, getters }, message) {
+      // only add if messages doesn't exist yet
 
-  // receive initial messages
-  [types.REQUEST_MESSAGES] (state, { conversationId }) {
-    state.receiveStatus = {
-      isWaiting: true,
-      error: null,
-      success: false,
-    }
-  },
-  [types.RECEIVE_MESSAGES] (state, { conversationId, messages, cursor }) {
-    if (state.messages[conversationId]) {
-      Vue.set(state.messages, conversationId, messages)
-      Vue.set(state.cursors, conversationId, cursor)
-    }
-    if (state.activeConversationId === conversationId) {
-      state.receiveStatus = {
-        isWaiting: false,
-        error: null,
-        success: true,
-      }
-    }
-  },
-  [types.RECEIVE_MESSAGES_ERROR] (state, { conversationId, error }) {
-    console.log(error)
-    if (state.activeConversationId === conversationId) {
-      state.receiveStatus = {
-        isWaiting: false,
-        error,
-        success: false,
+      if (!getters.activeMessages.find(e => e.id === message.id)) {
```

Comment:
A bit inefficient to go through all the messages each time, ok for now I'm sure but could be much improved.

Reply:
I agree, we could map the messages by ID, store in an object and then do a simple key lookup. But I'm sure that this is not the reason why the page is sluggish sometimes ;)

Reply:
I was thinking about an  "always sorted list" approach (would also fit with de-duplication)